ethereum/eth/src/main/java/org/hyperledger/besu/ethereum/eth/sync/fastsync/FastSyncTargetManager.java

```diff
@@ -167,8 +167,16 @@ private boolean peerHasDifferentPivotBlock(final List<BlockHeader> result) {
   @Override
   public boolean shouldContinueDownloading() {
     final BlockHeader pivotBlockHeader = fastSyncState.getPivotBlockHeader().get();
-    return !protocolContext.getBlockchain().getChainHeadHash().equals(pivotBlockHeader.getHash())
-        || !worldStateStorage.isWorldStateAvailable(
-            pivotBlockHeader.getStateRoot(), pivotBlockHeader.getBlockHash());
+    boolean isValidChainHead =
+        protocolContext.getBlockchain().getChainHeadHash().equals(pivotBlockHeader.getHash());
+    if (!isValidChainHead) {
+      if (protocolContext.getBlockchain().contains(pivotBlockHeader.getHash())) {
+        protocolContext.getBlockchain().rewindToBlock(pivotBlockHeader.getHash());
```

Comment:
Could the pivot block be > the max allowable bonsai trie layers away? If so, could this rewind take an unfeasible amount of time or maybe error? If so, then what state is the user left in and is it recoverable?

Reply:
this part rewind only the blockchain part not the worldstate part. 
If the pivot block is before the current head, we do a rewind otherwise we do nothing. In any case the worldstate will be healed with the new pivot block without doing a rollback or rollforward because we cannot do it if it is corrupted

it is important to choose a new pivot block and not to heal the current head. Because we are not sure that block will remain on the canonical chain